# Module 5 Assignment: Functions & Modularity
**CS 5045: Computation for the Data Sciences**

**Name:** Brian A Thomas

---

Complete each section below by filling in the blanks marked `__________` or writing code where indicated. Run all cells top to bottom and make sure there are no errors. Commit and push this notebook (`hw5.ipynb`) to your fork on GitHub, no Canvas upload needed.

**Dataset:** `data.csv`: road race results, 16 runners. The file is in this same folder.

In [7]:
#!pip install pandas numpy matplotlib seaborn

import pandas as pd

# 3. LOAD YOUR DATASET USING YOUR EXACT WINDOWS ABSOLUTE PATH
# (Using the 'r' prefix prevents the backslashes from breaking the code)
data_path = r'C:\Users\Powerzone.aafes\PycharmProjects\cs5045_BU\seminar_2\data (1).csv'
df = pd.read_csv(data_path)

# Print the first few rows to verify it successfully loaded
df.head()



,runner_id,name,finish_time_min,age,gender,category
0,R001,Amara Osei,42.3,29,F,Open
1,R002,Brian Tran,38.7,34,M,Open
2,R003,Celia Moreno,51.2,47,F,Masters
3,R004,David Park,44.8,31,M,Open
4,R005,Fatima Al-Hassan,56.4,52,F,Masters


## Section 1: Define a Transform function

In [10]:
# 1a. Complete the function below.
#     It should compute pace as finish time divided by distance, rounded to 2
#     decimal places. The default race distance is 10.0 km.

def pace_per_km(finish_time_min, distance_km=10.0):
    """
    TRANSFORM: compute pace in minutes per kilometer.

    Parameters:
        finish_time_min (float) -- runner's finish time in minutes
        distance_km (float)     -- race distance in km (default: 10.0)

    Returns:
        float -- pace in minutes per km, rounded to 2 decimal places
    """
    return round(finish_time_min/distance_km, 2)


# 1b. Call pace_per_km with just a finish time (default distance).
#     Store the result in pace_default.
pace_default = pace_per_km(42.3)
print("Default pace (10 km):", pace_default)
# Expected output: 4.23

# 1c. Call pace_per_km with distance_km=5.0.
#     Store the result in pace_5k.
pace_5k = pace_per_km(42.3, distance_km=5.0)
print("Pace for 5 km race:", pace_5k)
# Expected output: 8.46

Default pace (10 km): 4.23
Pace for 5 km race: 8.46


## Section 2: Parameters and return values

In [25]:
# 2a. Complete the age_group function below.
#     It should return:
#       "Under 30" if age < 30
#       "30-39"    if 30 <= age <= 39
#       "40-49"    if 40 <= age <= 49
#       "50+"      if age >= 50

def age_group(age):
    """
    TRANSFORM: classify a runner's age into a group string.

    Parameters:
        age (int) -- runner's age in years

    Returns:
        str -- age group label
    """
    if age < 30:
        return "Under 30"
    elif age <= 39:
        return "30-39"
    elif age <= 49:
        return "40-49"
    else:
        return "50+"


# 2b. Test your function with the ages below.
print("Age group tests:")
print(age_group(22))   # Expected: Under 30
print(age_group(34))   # Expected: 30-39
print(age_group(47))   # Expected: 40-49
print(age_group(58))   # Expected: 50+

# 2c. Complete the is_podium function below.
#     It should return True if finish_time_min is less than or equal to cutoff,
#     and False otherwise. Default cutoff is 41.0 minutes.

def is_podium(finish_time_min, cutoff=41.0):
    """
    FILTER: determine whether a finish time qualifies for the podium.

    Parameters:
        finish_time_min (float) -- runner's finish time in minutes
        cutoff (float)          -- maximum qualifying time (default: 41.0)

    Returns:
        bool -- True if finish_time_min <= cutoff, else False
    """
    return finish_time_min <= cutoff


print("\nPodium tests:")
print(is_podium(35.6))   # Expected: True
print(is_podium(42.3))   # Expected: False
print(is_podium(40.0))   # Expected: True

Age group tests:
Under 30
30-39
40-49
50+

Podium tests:
True
False
True


## Section 3: Apply functions to a DataFrame column

In [28]:
# 3a. Add a "pace_min_per_km" column to df using .apply() and pace_per_km.
df["pace_min_per_km"] = df['finish_time_min'].apply(pace_per_km)

# 3b. Add an "age_group" column to df using .apply() and age_group.
df["age_group"] = df['age'].apply(age_group)

# 3c. Add an "is_podium" column to df using .apply() and is_podium.
df["is_podium"] = df['is_podium'].apply(is_podium)

print("First 5 rows with new columns:")
print(df[["name", "finish_time_min", "pace_min_per_km", "age_group", "is_podium"]].head())

# 3d. How many runners qualify for the podium?
#     Filter df to rows where is_podium is True and print the count.
podium_count = df['is_podium'].sum()
print("\nRunners on the podium:", podium_count)
# Expected: 4  (Idris Okafor 35.6, Marcus Webb 37.2, Brian Tran 38.7, George Kim 40.1)

First 5 rows with new columns:
               name  finish_time_min  pace_min_per_km age_group  is_podium
0        Amara Osei             42.3             4.23  Under 30       True
1        Brian Tran             38.7             3.87     30-39       True
2      Celia Moreno             51.2             5.12     40-49       True
3        David Park             44.8             4.48     30-39       True
4  Fatima Al-Hassan             56.4             5.64       50+       True

Runners on the podium: 16


## Section 4: Refactor a repeated block into a function

Below is a repeated calculation written inline twice. Your task: wrap it in a function named `summarize_category`, then call it.

In [30]:
# Repeated version (do not change these lines):
masters = df[df["category"] == "Masters"]
masters_mean = round(masters["finish_time_min"].mean(), 2)
masters_min  = masters["finish_time_min"].min()
masters_max  = masters["finish_time_min"].max()

open_runners = df[df["category"] == "Open"]
open_mean = round(open_runners["finish_time_min"].mean(), 2)
open_min  = open_runners["finish_time_min"].min()
open_max  = open_runners["finish_time_min"].max()


def summarize_category(dataframe, category_name):
    """
    Computes the mean, minimum, and maximum finish times for a specific runner category.

    Parameters:
        dataframe (DataFrame) -- The dataset containing runner records
        category_name (str)   -- The specific category to filter by (e.g., "Masters")

    Returns:
        tuple -- (mean_time, min_time, max_time)
    """
    # Filter rows matching the specified category string- I used AI to help with this
    filtered_df = dataframe[dataframe["category"] == category_name]

    # Compute the requested aggregation metrics
    category_mean = round(filtered_df["finish_time_min"].mean(), 2)
    category_min  = filtered_df["finish_time_min"].min()
    category_max  = filtered_df["finish_time_min"].max()

    return category_mean, category_min, category_max


# Call the function for the "Masters" category-- I used AI to help with this
masters_mean, masters_min, masters_max = summarize_category(df, "Masters")
print(f"Masters Summary - Mean: {masters_mean}, Min: {masters_min}, Max: {masters_max}")

# Call the function for the "Open" category-- I used AI to help with this
open_mean, open_min, open_max = summarize_category(df, "Open")
print(f"Open Summary - Mean: {open_mean}, Min: {open_min}, Max: {open_max}")



Masters Summary - Mean: 54.0, Min: 48.9, Max: 61.0
Open Summary - Mean: 41.95, Min: 35.6, Max: 49.4


In [31]:
# 4a. Write a function summarize_category(dataframe, category_name) that:
#       - filters the dataframe to rows matching category_name
#       - returns a dict with keys "mean", "min", "max"
#     The mean should be rounded to 2 decimal places.

def summarize_category(dataframe, category_name):
    """
    AGGREGATE: compute mean, min, and max finish time for one category.

    Parameters:
        dataframe (DataFrame) -- the full race results DataFrame
        category_name (str)   -- the category to filter for ("Open" or "Masters")

    Returns:
        dict -- {"mean": float, "min": float, "max": float}
    """
    group = dataframe[dataframe["category"] == category_name]
    return {
        "mean": float(round(group['finish_time_min'].mean(), 2)),
        "min":  float(group['finish_time_min'].min()),
        "max":  float(group['finish_time_min'].max()),
    }


# 4b. Call summarize_category for each category and print the results.
masters_summary = summarize_category(df, "Masters")
open_summary    = summarize_category(df, "Open")

print("Masters summary:", masters_summary)
# Expected: {'mean': 54.0, 'min': 48.9, 'max': 61.0}

print("Open summary:   ", open_summary)
# Expected: {'mean': 41.95, 'min': 35.6, 'max': 49.4}

Masters summary: {'mean': 54.0, 'min': 48.9, 'max': 61.0}
Open summary:    {'mean': 41.95, 'min': 35.6, 'max': 49.4}


## Section 5: Write a function from scratch

Write a complete function called `fastest_in_category`. Do not fill in blanks; write the entire function yourself: the `def` line, a docstring, the body, and the return statement.

**Specification:**

- **Parameters:** `df` (the full race results DataFrame) and `category` (a string, either `"Open"` or `"Masters"`)
- **Returns:** a dictionary with two keys: `"name"` (the full name of the fastest runner in that category) and `"time"` (their finish time in minutes)
- The fastest runner is the one with the **minimum** finish time in that category.

Once you have written the function, call it for both categories and print each result.

In [34]:
# 5. Write fastest_in_category below.
#    Write the complete function: def line, docstring, body, and return.
#    Parameters: df (DataFrame), category (str)
#    Returns: dict with keys "name" (str) and "time" (float)




# Call fastest_in_category for both categories and print the results.
# Expected output:
#   Open:    {'name': 'Idris Okafor', 'time': 35.6}

def fastest_in_category(df, category):
    """
    Finds the fastest runner and their finish time within a specific category.

    Parameters:
        df (DataFrame) -- The full race results DataFrame containing 'category',
                          'name', and 'finish_time_min' columns
        category (str) -- The runner category to isolate ("Open" or "Masters")

    Returns:
        dict -- A dictionary containing the keys "name" (str) and "time" (float)
    """
    # Filter the DataFrame to only include rows from the specified category
    category_df = df[df["category"] == category]

    # Locate the row index containing the absolute minimum finish time
    fastest_row_idx = category_df["finish_time_min"].idxmin()
    fastest_runner = category_df.loc[fastest_row_idx]

    # Return the name and time packed cleanly inside a dictionary
    return {
        "name": fastest_runner["name"],
        "time": float(fastest_runner["finish_time_min"])
    }



# Call and print for the Open category
fastest_open = fastest_in_category(df, "Open")
print("Fastest in Open category:", fastest_open)

# Call and print for the Masters category
fastest_masters = fastest_in_category(df, "Masters")
print("Fastest in Masters category:", fastest_masters)


Fastest in Open category: {'name': 'Idris Okafor', 'time': 35.6}
Fastest in Masters category: {'name': 'Hannah Schmidt', 'time': 48.9}


## Section 6: Read and describe

Below is a complete function with a docstring. **Do not change the code cell.** Read the code carefully, run the cell to see the output, then answer the three questions in the markdown cells below.

In [35]:
# Do not change this cell.

def category_gap(df):
    """
    AGGREGATE: compute the difference in mean finish time between
    Masters and Open runners, in minutes.

    Parameters:
        df (pd.DataFrame) -- the full race results DataFrame

    Returns:
        float -- mean finish time of Masters runners minus mean finish time
                 of Open runners, rounded to two decimal places
    """
    open_mean = df[df["category"] == "Open"]["finish_time_min"].mean()
    masters_mean = df[df["category"] == "Masters"]["finish_time_min"].mean()
    return round(masters_mean - open_mean, 2)


# Run this to see the result.
print("Category gap:", category_gap(df))

Category gap: 12.05


**6a.** Describe what `category_gap` does in 2–3 sentences, as if explaining to a race director who does not code. Do not reference Python syntax - explain what it computes and what the result tells you.

**Your answer:** The category gap here is comparing the two runners division or groups ( open and Masters).  It is showing a comparison to illustrate some type of performance of each groups verses the individuals. It is looking at pace of each group as whole. The section before this look at individuals pace.

**6b.** For this dataset, `category_gap` returns a positive number. What does that tell you about how the Masters and Open runners compare?

**Your answer:** This positive number tells us the Master group is +12.05 slower , on average, than the Open runner.

**6c.** What would a return value of `0.0` mean, in plain English?

**Your answer:** The 0.0 would indicate there is no performance gap between the two groups (open and master).  It is possible there may be no differance in individuals, but when looking at the overall averages for the groups, there is a gap.